src/Core/Datatypes/DyadicTensor.h

```diff
@@ -0,0 +1,306 @@
+        return eigvecs_[index];
+      }
+
+      std::vector<DenseColumnMatrixGeneric<Number>> getEigenvectors()
+      {
+        if (!haveEigens_) buildEigens();
+        return eigvecs_;
+      }
+
+      std::vector<Number> getEigenvalues()
+      {
+        if (!haveEigens_) buildEigens();
+        return eigvals_;
+      }
+
+      SCISHARE friend std::ostream& operator<<(
+          std::ostream& os, const DyadicTensorGeneric<Number, Dim0, Dim1>& other)
+      {
+        os << '[';
+        for (size_t i = 0; i < other.getDimension1(); ++i)
+          for (size_t j = 0; j < other.getDimension2(); ++j)
+          {
+            if (i + j != 0) os << ' ';
+            os << other(other.index(j), other.index(i));
+          }
+        os << ']';
+
+        return os;
+      }
+
+      SCISHARE friend std::istream& operator>>(
+          std::istream& is, DyadicTensorGeneric<Number, Dim0, Dim1>& other)
+      {
+        char bracket;
+        is >> bracket;
+        for (size_t i = 0; i < other.getDimension1(); ++i)
+          for (size_t j = 0; j < other.getDimension2(); ++j)
+            is >> other(other.index(j), other.index(i));
+        is >> bracket;
+
+        return is;
+      }
+
+      DyadicTensorGeneric<Number, Dim0, Dim1>& operator=(const Number& v)
+      {
+        for (size_t i = 0; i < getDimension1(); ++i)
+          for (size_t j = 0; j < getDimension2(); ++j)
+            (*this)(index(i), index(j)) = v;
+        haveEigens_ = false;
+        return *this;
+      }
+
+      size_t getDimension1() const { return Dim0; }
+      size_t getDimension2() const { return Dim1; }
+
+      Number frobeniusNorm()
+      {
+        if (!haveEigens_) buildEigens();
+        auto eigvals = DenseColumnMatrix(eigvals_);
+        return eigvals.norm();
+      }
+
+      Number maxNorm()
+      {
+        if (!haveEigens_) buildEigens();
+        auto maxVal = std::numeric_limits<Number>::min();
+        for (const auto& e : eigvals_)
+          maxVal = std::max(maxVal, e);
+        return maxVal;
+      }
+
+      void setEigens(
+          const std::vector<DenseColumnMatrix>& eigvecs, const std::vector<Number>& eigvals)
+      {
+        if (eigvecs_.size() != eigvecs.size())
+          THROW_INVALID_ARGUMENT("The number of input eigvecs must be " + eigvecs_.size());
+        if (eigvals_.size() != eigvals.size())
+          THROW_INVALID_ARGUMENT("The number of input eigvals must be " + eigvals_.size());
+        eigvecs_ = eigvecs;
+        eigvals_ = eigvals;
+        haveEigens_ = true;
+      }
+
+      Number eigenValueSum()
+      {
+        if (!haveEigens_) buildEigens();
+        Number sum = 0;
+        for (const auto& e : eigvals_)
+          sum += e;
+        return sum;
+      }
+
+     protected:
+      const int RANK_ = 2;
+      std::vector<DenseColumnMatrixGeneric<Number>> eigvecs_;
+      std::vector<Number> eigvals_;
+      bool haveEigens_ = false;
+
+      void buildEigens()
```

Comment:
Consider making `eigvals_` and `haveEigens_` mutable, so that almost every function here can be marked const, assuming they don't change the underlying matrix/tensor values.